plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+	for (i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio (device, i, ptr, error))
+			return FALSE; /* FIXME also print a error message */
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+/*
+ * The mailbox can hold 64 bytes in 16 doubleword data registers.
+ * A NVM write operation writes data from these registers to NVM
+ * at the set offset
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_write (FuDevice *device, const guchar *data,
+				   guint8 length, GError **error)
+{
+	guchar *ptr = (guchar *) data;
+
+	if (length > 64 || length % 4) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "invalid firmware data write length %u\n",
+			     length);
+		return FALSE;
+	}
+	/* writes 4 bytes per iteration */
+	for (gint i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_set_mmio (device, i, ptr, error))
+			return FALSE;
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_operation (FuDevice *device, guint16 opcode,
+				 guchar *metadata, GError **error)
+{
+	g_autoptr (GError) error_local = NULL;
+	struct mbox_regx *regx;
+	guchar buf[4];
+	gint   i;
+
+	regx = (struct mbox_regx *)buf;
+	regx->opcode = GUINT16_TO_LE(opcode);
+	regx->status = MBOX_OPVALID;
+
+	/* Write metadata register for operations that use it */
+	switch (opcode) {
+		case OP_NVM_WRITE:
+		case OP_NVM_AUTH_WRITE:
+			break;
+		case OP_NVM_READ:
+		case OP_NVM_SET_OFFSET:
+		case OP_DROM_READ:
+			if (!metadata) {
+				g_set_error (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_INVALID_DATA,
+					     "hub opcode 0x%x requires metadata",
+					     opcode);
+				return FALSE;
+			}
+			if (!fu_dell_dock_usb4_hub_set_mmio (device,
+							     MBOX_REG_METADATA,
+							     metadata, error)) {
+				g_prefix_error (error,
+						"failed to write metadata %s:",
+						metadata);
+				return FALSE;
+			}
+			break;
+		default:
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "invalid hub opcode: 0x%x", opcode);
+			return FALSE;
+	}
+
+	/* write the operation and poll completion or error */
+	if (!fu_dell_dock_usb4_hub_set_mmio (device, MBOX_REG, buf, error))
+		return FALSE;
+
+	for (i = 0; i < 100; i++) {
+		g_clear_error (&error_local);
+		if (fu_dell_dock_usb4_hub_get_mmio (device, MBOX_REG, buf, &error_local))
+			return TRUE;
+
+		g_usleep ((gulong) 10000);
+	}
+
+	g_set_error (error, G_IO_ERROR, G_IO_ERROR_TIMED_OUT,
+		     "maximum tries exceeded: %s", error_local->message);
+	return FALSE;
+}
+
+/**
+ * fu_dell_dock_usb4_hub_nvm_read:
+ * @device: a #FuDevice
+ * @buf: array of bytes to store the read from registers
+ * @length: length of buf array
+ * @nvm_addr: nvm read offset, e.g. `0`
+ * @error: (nullable): optional return location for an error
+ *
+ * Read NVM over USB interface
+**/
+static gboolean
+fu_dell_dock_usb4_hub_nvm_read (FuDevice *device, guint8 *buf, guint32 length,
+			        guint32 nvm_addr, GError **error)
+{
+	guint8 tmpbuf[64];
```

Comment:
I would feel better if this was initialized, so that if there was a bug and `fu_dell_dock_usb4_mbox_data_read` didn't populate it, it's got zeros instead of garbage.